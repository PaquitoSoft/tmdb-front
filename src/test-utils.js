import React from 'react';
import { shallow, mount } from 'enzyme';
import { AppProvider } from './components/app-context/app-context';

const userIdMock = 'hskfjahshjdjskla';

const imagesConfigMock = {
	"baseUrl":"https://image.tmdb.org/t/p/",
	"backdropSizes":["w300","w780","w1280","original"],
	"logoSizes":["w45","w92","w154","w185","w300","w500","original"],
	"posterSizes":["w92","w154","w185","w342","w500","w780","original"],
	"profileSizes":["w45","w185","h632","original"],
	"stillSizes":["w92","w185","w300","original"]
};

const renderers = {
	'shallow': shallow,
	'mount': mount
};

function AppContextMock({ children, apiClientMock }) {
	return (
		<AppProvider 
			apiClient={apiClientMock} 
			userId={userIdMock} 
			imagesConfig={imagesConfigMock}
		>
			{children}
		</AppProvider>
	);
}

export function render(
	component, 
	options = { renderType: 'shallow' }
) {
	return renderers[options.renderType](component);
}

export function validateSnapshot(
	component, 
	options = { renderType: 'shallow' }
) {
	const wrapper = render(component, options);
	expect(wrapper).toMatchSnapshot();
	return wrapper;
}

export function renderInAppContext(
	component, 
	options = {}
) {
	const _options = {
		renderType: 'mount',
		apiClientQueryMock: jest.fn().mockImplementation(() => Promise.resolve({})),
		apiClientMutationMock: jest.fn().mockImplementation(() => Promise.resolve({})),
		...options
	};

	const apiClientMock = {
		sendQuery: _options.apiClientQueryMock,
		sendMutation: _options.apiClientMutationMock
	};
	
	const wrapper = render(
		(
			<AppContextMock apiClientMock={apiClientMock}>
				{component}
			</AppContextMock>
		),
		_options
	);

	return wrapper.find(AppProvider).childAt(0);
}

export const clone = object => JSON.parse(JSON.stringify(object));
