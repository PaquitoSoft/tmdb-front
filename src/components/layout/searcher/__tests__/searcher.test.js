import React from 'react';
import { act } from 'react-dom/test-utils';
import { validateSnapshot, renderInAppContext } from '../../../../test-utils';

import Searcher from '../searcher';

import searchResultMock from './search-results-fixture.json';

describe('Searcher', () => {

	it('Should render Searcher component', () => {
		validateSnapshot(<Searcher />);
	});

	it('Should call server when submitting a search term', () => {
		const searchTerm = 'Heroes';
		const queryMock = jest.fn().mockImplementation(
			() => Promise.resolve({ data: searchResultMock })
		);
		const wrapper = renderInAppContext(
			<Searcher />,
			{
				apiClientQueryMock: queryMock
			}
		);
		
		expect(wrapper.exists('.searcher__results')).toEqual(false);

		wrapper
			.find('.searcher__search-term')
			.simulate('focus')
			.simulate('change', {
				target: { value: searchTerm }
			});
		
		act(() => {
			wrapper
				.find('form')
				.simulate('submit', {
					preventDefault: () => false
				});
		});
		
		const { query, variables } = queryMock.mock.calls[0][0];
		expect(query).toEqual(expect.stringContaining('searchTvShows'));
		expect(variables.searchTerm).toEqual(searchTerm);
		expect(wrapper.exists('.searcher__results')).toEqual(true);

	});

	it('Should show results panel when server returns results', () => {
		throw new Error('Not yet implemented');
	});

	it('Should hide results panel when clicking outside Searcher component', () => {
		throw new Error('Not yet implemented');
	});

});
