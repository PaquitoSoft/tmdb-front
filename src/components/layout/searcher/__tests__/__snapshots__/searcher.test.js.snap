// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searcher Should render Searcher component 1`] = `
<div
  className="searcher"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="searcher__search-term"
      name="searchTerm"
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Type in your search term"
      type="text"
      value=""
    />
  </form>
</div>
`;
